packages/terminal/src/browser/search/terminal-search-widget.tsx

```diff
@@ -0,0 +1,173 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from 'inversify';
+import { ReactWidget } from '@theia/core/lib/browser/widgets/react-widget';
+import * as React from 'react';
+import '../../../src/browser/search/terminal-search.css';
+import { Terminal } from 'xterm';
+import * as ReactDOM from 'react-dom';
+import { findNext, findPrevious } from 'xterm/lib/addons/search/search';
+import { ISearchOptions } from 'xterm/lib/addons/search/Interfaces';
+import { Key } from '@theia/core/lib/browser';
+
+export const TerminalSearchWidgetFactory = Symbol('TerminalSearchWidgetFactory');
+export type TerminalSearchWidgetFactory = (terminal: Terminal, node: Element, terminalWdgId: string) => TerminalSearchWidget;
+
+export enum TerminalSearchOption {
+    CaseSensitiv = 'caseSensitive',
+    WholeWord = 'wholeWord',
+    RegExp = 'regex'
+}
+
+@injectable()
+export class TerminalSearchWidget extends ReactWidget {
+
+    private searchInput: HTMLInputElement | null;
+    private searchBox: HTMLDivElement | null;
+    private searchOptions: ISearchOptions = {};
+
+    @inject(Terminal)
+    protected terminal: Terminal;
+
+    @inject(Element)
+    protected element: Element;
+
+    @postConstruct()
+    protected init() {
+        this.hide();
+        this.element.appendChild(this.node);
+        ReactDOM.render(<React.Fragment>{this.render()}</React.Fragment>, this.node);
+    }
+
+    isActivated(): boolean {
+        return this.node.clientWidth > 0;
+    }
+
+    focus(): void {
+        if (this.searchInput) {
+            this.searchInput.focus();
+        }
+    }
+
+    update(): void {
+        ReactDOM.render(<React.Fragment>{this.render()}</React.Fragment>, this.node);
+    }
+
+    render(): React.ReactNode {
+        this.node.classList.add('find-terminal-widget-parent');
+        return <div className='find-terminal-widget'>
+            <div className='search-elem-box' ref={searchBox => this.searchBox = searchBox} >
+                <input
+                    title='Find'
+                    type='text'
+                    placeholder='Find'
+                    ref={ip => this.searchInput = ip}
+                    onKeyUp={event => this.search(event)}
+                    onFocus={() => this.onSearchInputFocus()}
```

Comment:
never use arrow functions for event handlers, it breaks react cachgin, use following pattern:
```ts
onKeyUp={this.search}

protected search = (event: React.KeyboardEvent) => {
}
```

It creates a single instance of a function with `this` access and passes it to React by a reference for caching.

Reply:
Added it to the coding guideline: https://github.com/theia-ide/theia/wiki/Coding-Guidelines#no-bind-fn-in-event-handlers